File: ocaml/d2-1.ml
let rec process_lines () = 
  try
    let line = read_line () in
    let a = String.get line 0 in
    let b = String.get line 2 in
    process_lines () + match b with
    | 'X' -> 1 + (
      match a with
      | 'A' -> 3
      | 'B' -> 0
      | 'C' -> 6
      | _ -> 0)
    | 'Y' -> 2 + (
      match a with
      | 'A' -> 6
      | 'B' -> 3
      | 'C' -> 0
      | _ -> 0)
    | 'Z' -> 3 + (
      match a with
      | 'A' -> 0
      | 'B' -> 6
      | 'C' -> 3
      | _ -> 0)
    | _ -> 0
  with
  _ -> 0;;
  
print_int (process_lines ());;

File: rust/d2-1.rs
use std::io::{self, BufRead};

fn main() -> io::Result<()> {
    let lines = io::stdin().lock().lines();
    let mut score = 0;
    for line in lines {
        let line = line?;
        let mut chars = line.chars();
        let a = chars.nth(0);
        let b = chars.nth(1);
        match b.unwrap() {
            'X' => {
                score += 1;
                match a.unwrap() {
                    'A' =>  score += 3,
                    'B' =>  score += 0,
                    'C' =>  score += 6,
                    _ => println!("Invalid character"),
                }
            },
            'Y' => {
                score += 2;
                match a.unwrap() {
                    'A' =>  score += 6,
                    'B' =>  score += 3,
                    'C' =>  score += 0,
                    _ => println!("Invalid character"),
                }
            },
            'Z' => {
                score += 3;
                match a.unwrap() {
                    'A' =>  score += 0,
                    'B' =>  score += 6,
                    'C' =>  score += 3,
                    _ => println!("Invalid character"),
                }
            },
            _ => println!("Invalid character"),
        }
    }
    println!("{}", score);
    Ok(())
}
